Async batch

vzhestkov opened the pull request:
### What does this PR do?

Implementation of [RFC#0002](https://github.com/saltstack/salt/blob/develop/rfcs/0002-netapi-async-batch.md)

https://github.com/saltstack/salt/pull/50546 port to `master`

All related tests migrated to `pytest`

vzhestkov committed:
Async batch implementation RFC#0002

Co-authored-by: Mihai Dinca <dincamihai@gmail.com>

garethgreenaway commented:
@vzhestkov Just a heads up, there are quite a few merge conflicts here.

vzhestkov commented:
@garethgreenaway thanks for pinging, I'll resolve them once I get some time for it, will try to do it as soon as possible

twangboy commented on salt/cli/batch.py:
```diff
@@ -13,9 +16,89 @@
 
+def get_bnum(opts, minions, quiet):
+    """
+    Return the active number of minions to maintain
```
Add Args and Return to docs
Also needs `.. versionadded:: Phosphorus`

twangboy commented on salt/cli/batch.py:
```diff
@@ -13,9 +16,89 @@
+
+def batch_get_opts(
+    tgt, fun, batch, parent_opts, arg=(), tgt_type="glob", ret="", kwarg=None, **kwargs
+):
```
Needs docs
Add Args and Return to docs
Also needs `.. versionadded:: Phosphorus`

twangboy commented on salt/cli/batch.py:
```diff
@@ -13,9 +16,89 @@
+    return opts
+
+
+def batch_get_eauth(kwargs):
```
Docs
Add Args and Return to docs
Also needs `.. versionadded:: Phosphorus`

twangboy commented on salt/cli/batch.py:
```diff
@@ -74,25 +162,7 @@ def __gather_minions(self):
-                    "Invalid batch data sent: {}\nData must be in the "
-                    "form of %10, 10% or 3".format(self.opts["batch"])
-                )
+        return get_bnum(self.opts, self.minions, self.quiet)
```
Won't this just call itself... recursively?

vzhestkov replied:
It's not calling `get_bnum` from `self`, so it will call `get_bnum` from global.

twangboy commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,315 @@
+"""
+Execute a job on the targeted minions by using a moving window of fixed size `batch`.
```
Needs `.. versionadded:: Phosphorus`

twangboy commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,315 @@
+        )
+        return set(list(to_run)[:next_batch_size])
+
+    def check_find_job(self, batch_minions, jid):
```
Could we get some docs here?

twangboy commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,315 @@
+                self.event.io_loop.spawn_callback(self.find_job, running)
+
+    @salt.ext.tornado.gen.coroutine
+    def find_job(self, minions):
```
Docs

twangboy commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,315 @@
+                self.close_safe()
+
+    @salt.ext.tornado.gen.coroutine
+    def start(self):
```
Docs

twangboy commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,315 @@
+                self.event.io_loop.spawn_callback(self.start_batch)
+
+    @salt.ext.tornado.gen.coroutine
+    def start_batch(self):
```
Docs

twangboy commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,315 @@
+                self.event.io_loop.spawn_callback(self.run_next)
+
+    @salt.ext.tornado.gen.coroutine
+    def end_batch(self):
```
Docs

Ch3LL commented:
There are test failures

garethgreenaway commented:
@vzhestkov A few test failures here and some requested changes.

vzhestkov commented:
Thanks for pinging. I'll try to fix all the suggestions today or tomorrow morning

twangboy commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,340 @@
+        - gather_job_timeout: `find_job` timeout
+        - timeout: time to wait before firing a `find_job`
+
+    When the batch stars, a `start` event is fired:
```
When the batch starts

twangboy commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,340 @@
+             "down_minions": targeted_minions - presence_ping_minions
+           }
+
+    When the batch ends, an `done` event is fired:
```
a `done` event

twangboy commented on salt/master.py:
```diff
@@ -2163,6 +2165,22 @@ def get_token(self, clear_load):
             return False
         return self.loadauth.get_tok(clear_load["token"])
 
+    def publish_batch(self, clear_load, minions, missing):
```
Still need some docs here, please.

twangboy replied:
And this is a new function, so it will need a `.. versionadded:: 3005` as well

twangboy commented on salt/utils/event.py:
```diff
@@ -920,6 +920,10 @@ def fire_ret_load(self, load):
                     # Minion fired a bad retcode, fire an event
                     self._fire_ret_load_specific_fun(load)
 
+    def remove_event_handler(self, event_handler):
```
Please add some docs and a `.. versionadded: 3005`

vzhestkov commented:
@twangboy thanks for pointing again, accidentally missed it. Fixed

Ch3LL commented:
One thing that is really helpful is to check the console output and copy/paste the exact nox command that is run: `nox -f /tmp/kitchen/testing/noxfile.py  -e 'pytest-zeromq-m2crypto-3(coverage=True)'  -- --output-columns=120 --sys-stats --sysinfo --junitxml=/tmp/kitchen/testing/artifacts/xml-unittests-output/test-results-20220617090644.507.xml -vv --run-destructive --ssh-tests --run-slow --from-filenames=salt/auth/__init__.py,salt/cli/batch.py,salt/cli/batch_async.py,salt/client/__init__.py,salt/master.py,salt/transport/ipc.py,salt/utils/event.py,tests/pytests/unit/cli/test_batch.py,tests/pytests/unit/cli/test_batch_async.py,tests/unit/cli/test_daemons.py,tests/unit/cli/__init__.py,tests/unit/cli/test_batch.py,tests/unit/transport/test_ipc.py`

The documentation for the test suite is here: https://docs.saltproject.io/en/latest/topics/tutorials/writing_tests.html

vzhestkov commented:
@Ch3LL I've moved the failing test to more relevant place with readjusting the code, but there are some other tests failing for different OSes for some reasons, the fails don't look related.

@waynew could you please take a look. I don't understand why this test is failing: `test_when_min_wait_time_has_not_passed_then_disbatch_should_not_return_expected_data_until_time_has_passed`
I can't reproduce this fail with running this test individually on local instance.

garethgreenaway commented:
@vzhestkov Looks like there are some merge conflicts on this one.

MKLeb commented:
@vzhestkov it looks like there are some related failing tests for this one.

vzhestkov commented:
Hi @MKLeb thanks for pinging. There was a comment about this failing test https://github.com/saltstack/salt/pull/60269#issuecomment-1239083255. I'm not sure it's related and it's not failing if run it individually. Maybe moving it to pytest will help, I can try, but not sure if it's the best way. I don't understand how this PR can affect this test.

vzhestkov commented:
@MKLeb I've fixed the failing test. Sorry I've missed the wrong mockings in the original tests created long time ago.

vzhestkov committed:
Use asyncio

vzhestkov committed:
Fix tornado usage

vzhestkov committed:
Fix async tests

vzhestkov committed:
Fix connected check

vzhestkov committed:
Remove remove_event_handler

vzhestkov committed:
Fix batch call

vzhestkov committed:
Fix async tests

meaksh committed:
Make batch async working with asyncio

batch async: enhance trace logging

batch async: do not block event handler

additionally, only try to execute next batch once all previous minions
has been processed somehow

batch async: do not block current find_job execution

batch async: just start batch if all minions returns the ping

batch_async: ensure the start/done events are fired

batch async: do not wait to finalize batch execution

batch async: only trigger find job for current batch active minions

Prevent traceback due missing 'send' method when socket was closed

Modify some batch_async trace logs

vzhestkov committed:
Make pre-commit check happy

twangboy commented:
Looks like there are some test failures here

twangboy commented:
Please fix the conflicts

dwoz commented:
@vzhestkov Can you please rebase this :pray: . We'll try to get it into 3008.0

vzhestkov committed:
Fix batch_async test after alignment with asyncio

dwoz committed:
Fix test_gather_minions_with_batch_presence_ping KeyError and add docs

dwoz commented:
Rebased onto master and addressed review feedback:

- **CI fix**: `gather_minions()` fallback from `self.opts["timeout"]` to `self.opts.get("timeout")` — prevented `KeyError` when `batch_presence_ping_timeout` is absent and `timeout` is not in opts
- **Test fix**: added missing `"timeout": 10` to `opts_no_pp` in `test_gather_minions_with_batch_presence_ping` (same test asserted `opts_no_pp["timeout"]` at the assertion site, which would also have raised `KeyError`)
- **Docs**: added `.. versionadded:: 3009.0` to `get_bnum`, `batch_get_opts`, `batch_get_eauth` (batch.py) and `BatchAsync` class (batch_async.py); added Args/Returns to `publish_batch` in master.py
- **Changelog**: added `changelog/60269.added.md`
- **Milestone**: updated to Potassium v3009.0

The PR is not superseded by #69444. That PR fixed a 3008.x sync batch regression; this PR adds the new async batch subsystem (`salt/cli/batch_async.py`) implementing RFC-0002, which doesn't exist on master yet.

dwoz committed:
Fix asyncio.timeout compatibility with Python < 3.11

asyncio.timeout() was added in Python 3.11. Replace with asyncio.wait_for()
which is compatible with Python 3.8+ as required by salt.
